In [46]:
import os
from PIL import Image
import numpy as np
import subprocess

In [47]:
dataset_path_root = "Data/medical data/"
datasets = ["melanoma_detection/", "skin_lesions_classification/"]

In [48]:
sum_rgb = np.zeros(3)
sum_sq_rgb = np.zeros(3)
num_pixels = 0

results = {}

#Loop trough type folders
for dataset in datasets:
    dataset_path = os.path.join(dataset_path_root, dataset)
    fileNames = []
    imgCount = 0
    for type_folder in os.listdir(dataset_path):
        type_folder_path = os.path.join(dataset_path, type_folder)
        if not os.path.isdir(type_folder_path): continue
        #Loop through class folders
        for class_folder in os.listdir(type_folder_path):
            class_folder_path = os.path.join(type_folder_path, class_folder)
            if not os.path.isdir(class_folder_path): continue
            #Loop through images
            for image_file in os.listdir(class_folder_path):
                image_path = os.path.join(class_folder_path, image_file)
                if not os.path.isfile(image_path): continue
                
                #Open image and convert to numpy array
                with Image.open(image_path) as img:
                    img_array = np.array(img.convert("RGB")) / 255.0
                    print(f"Image: {image_file}, Shape: {img_array.shape}, Dtype: {img_array.dtype}")

                    pixels = img_array.reshape(-1, 3)

                    sum_rgb += pixels.sum(axis=0)
                    sum_sq_rgb += (pixels ** 2).sum(axis=0)
                    num_pixels += pixels.shape[0]
                
                fileNames.append(image_file)
                imgCount += 1
                if imgCount % 100 == 0:
                    print(f"Processed {imgCount} images...")

    mean = sum_rgb / num_pixels
    std = np.sqrt((sum_sq_rgb / num_pixels) - (mean ** 2))


    occurances = {}
    duplicate_files = []
    for fName in fileNames:
        if fName not in occurances:
            occurances[fName] = 0
        occurances[fName] += 1

    for fName, count in occurances.items():
        if count > 1:
            print(f"File: {fName} occures more than once")
            duplicate_files.append(fName)

    
    results[dataset] = {}
    results[dataset]["mean"] = mean
    results[dataset]["std"] = std
    results[dataset]["duplicates"] = duplicate_files


Image: ISIC_0014790.jpg, Shape: (4441, 6641, 3), Dtype: float64
Image: ISIC_0014800.jpg, Shape: (4422, 6642, 3), Dtype: float64
Image: ISIC_0014826.jpg, Shape: (4401, 6621, 3), Dtype: float64
Image: ISIC_0014862.jpg, Shape: (4439, 6668, 3), Dtype: float64
Image: ISIC_0014872.jpg, Shape: (4459, 6668, 3), Dtype: float64
Image: ISIC_0014883.jpg, Shape: (4439, 6668, 3), Dtype: float64
Image: ISIC_0014912.jpg, Shape: (4439, 6708, 3), Dtype: float64
Image: ISIC_0014928.jpg, Shape: (4439, 6688, 3), Dtype: float64
Image: ISIC_0014932.jpg, Shape: (4439, 6688, 3), Dtype: float64
Image: ISIC_0014963.jpg, Shape: (4459, 6688, 3), Dtype: float64
Image: ISIC_0014982.jpg, Shape: (4420, 6641, 3), Dtype: float64
Image: ISIC_0015004.jpg, Shape: (4459, 6628, 3), Dtype: float64
Image: ISIC_0015041.jpg, Shape: (4402, 6621, 3), Dtype: float64
Image: ISIC_0015046.jpg, Shape: (4422, 6641, 3), Dtype: float64
Image: ISIC_0015050.jpg, Shape: (4382, 6601, 3), Dtype: float64
Image: ISIC_0015071.jpg, Shape: (4422, 6

In [49]:
print(results)

{'melanoma_detection/': {'mean': array([0.71074242, 0.54824915, 0.50507771]), 'std': array([0.16896467, 0.16984105, 0.18466681]), 'duplicates': []}, 'skin_lesions_classification/': {'mean': array([0.6750431 , 0.53589019, 0.50290742]), 'std': array([0.20735016, 0.19458835, 0.20519302]), 'duplicates': []}}


{'melanoma_detection/': {'mean': array([0.71074242, 0.54824915, 0.50507771]), 'std': array([0.16896467, 0.16984105, 0.18466681]), 'duplicates': []}, 'skin_lesions_classification/': {'mean': array([0.6750431 , 0.53589019, 0.50290742]), 'std': array([0.20735016, 0.19458835, 0.20519302]), 'duplicates': []}}

Melanome detection dataset:


Skin Lesions Classifications Dataset:



In [ ]:
#subprocess.run(["systemctl", "suspend"])

CompletedProcess(args=['systemctl', 'suspend'], returncode=0)